images/build/Dockerfiles/Dockerfile

```diff
@@ -1,40 +1,38 @@
-ARG IMAGES_DIR=/tmp/oryx/images
-ARG BUILD_DIR=/tmp/oryx/build
-# Start declaration of Build-Arg to determine where the image is getting built (DevOps agents or local)
-ARG AGENTBUILD
-ARG YARN_CACHE_BASE_TAG
-
 FROM oryxdevmcr.azurecr.io/public/oryx/build:lts-versions AS main
-ARG IMAGES_DIR
-ARG BUILD_DIR
-# End declaration of Build-Arg to determine where the image is getting built (DevOps agents or local)
 
-# Configure locale (required for Python)
-# NOTE: Do NOT move it from here as it could have global implications
-ENV LANG C.UTF-8
+###
+# Build run script generators (to be used by the `oryx run-script` command)
+###
+FROM golang:1.11-stretch as startupScriptGens
 
-# Install basic build tools
-RUN apt-get update \
-    && apt-get upgrade -y \
-    && apt-get install -y --no-install-recommends \
-        # Required for .NET Core 1.1
-        libunwind8 \
-    && rm -rf /var/lib/apt/lists/*
+# GOPATH is set to "/go" in the base image
+WORKDIR /go/src
+COPY src/startupscriptgenerator/src .
+
+ARG GIT_COMMIT=unspecified
+ARG BUILD_NUMBER=unspecified
+ARG RELEASE_TAG_NAME=unspecified
+ENV GIT_COMMIT=${GIT_COMMIT}
+ENV BUILD_NUMBER=${BUILD_NUMBER}
+ENV RELEASE_TAG_NAME=${RELEASE_TAG_NAME}
 
-# A temporary folder to hold all content temporarily used to build this image.
-# This folder is deleted in the final stage of building this image.
-RUN mkdir -p ${IMAGES_DIR}
-RUN mkdir -p ${BUILD_DIR}
-ADD build ${BUILD_DIR}
-ADD images ${IMAGES_DIR}
-RUN find ${IMAGES_DIR} -type f -iname "*.sh" -exec chmod +x {} \;
-RUN find ${BUILD_DIR} -type f -iname "*.sh" -exec chmod +x {} \;
+RUN ./build.sh dotnetcore /opt/startupcmdgen/dotnet
+RUN ./build.sh node       /opt/startupcmdgen/nodejs
+RUN ./build.sh php        /opt/startupcmdgen/php
+RUN ./build.sh python     /opt/startupcmdgen/python
+###
+# End build run script generators
+###
 
 # Install .NET Core
-FROM main AS dotnet-install
-ARG BUILD_DIR
-ARG IMAGES_DIR
-RUN apt-get update \
+FROM main AS pre-final
+COPY --from=tempfiles /tmp/oryx/ /opt/tmp
```

Comment:
Instead of multiple intermediate images (like dotnet-install, node-install), we have only 1 now and we try to copy all info to `/opt` folder because this allows us to use only 1 COPY instruction (hence only 1 layer) when copying from this intermediate stage to final stage. There is perf cost here but unavoidable in my opinion.